[Spring MVC(인증)]곰곰 미션 제출합니다. Step1

jaewoo9797 opened the pull request:
커찬 안녕하세요 👋 

인증 관련 미션을 진행하였습니다. 인증과 인가에 대해서 한 번 해보았지만 다시 해보니 기억이 잘 안나더군요.. 이번 step1 미션의 목표는 JWT 를 생성하고 파싱하는 것과 쿠키를 이용해서 인증하는 것이라고 생각했습니다.

## 쿠키 굽기 🍪 
쿠키를 만들어서 클라이언트에게 넘겨주는 방법은 `ResponseCookie` 객체에서 빌더를 리턴받아 생성하는 방식을 선택했습니다.
- `Cookie` 객체를 생성자를 통해서 생성하는 법.
- 처음 알았는데, 세션에 등록한 후에 별다른 조치 없이 응답해도 `Set-Cookie` 헤더에 세션 ID가 넘어가는 것

ResponseCookie 빌더를 이용하는것이 가장 가독성이 좋다고 생각했고, 설정하는 것도 명시적이라 선택했습니다. 그리고 ResponseEntity 헤더에 쿠키를 담아서 리턴합니다.

### 쿠키 Request에서 빼오기 
-  `@CookieValue` 
- WebUtils Class
- HttpServletRequest Class

어노테이션을 이용하여 쿠키를 Request에서 빼오는 방법을 선택했습니다. 다른 방법에 비해서 직관적이고, 간편했습니다.  null 에 대한 처리는 어떻게 되는지 찾아봐야합니다.

## JWT 토큰
현재 서비스 레이어에서 토큰을 생성 파싱하고 있습니다. 이를 분리하여 Jwt 토큰을 생성, 파싱하는 객체를 분리하기 전에 해당 로직이 서비스 레이어에서 분리되어 객체로 관리되어야 하는 이유에 대해서 생각하고 있었습니다. 

테스트 학습이나, 이전에 구현할 했을때는 JwtUtil , JwtProvider 등 객체의 책임을 가지도록 분리했습니다. 그런데 미션에서 그냥 그랬으니까, 그렇게 하는거니까로 분리하기가 싫어서 `왜 분리해야할까`에 대해서 고민하고 있습니다..

jaewoo9797 committed:
Docs 스터디를 소개할 수 있는 README.md 파일 생성

jaewoo9797 committed:
README 파일 이름 오타 수정

jaewoo9797 committed:
Create Login controller Logic

응답에 쿠키를 담는 방법에 대해서 알아보고 실제로 어떻게 나오는지 확인
쿠키를 만드는 방법에 대해서 다른 방법은 없는지 알아보기

jaewoo9797 committed:
Docs 응답 쿠키에 jwt 토큰을 담기 위해 주석으로 해야하는 부분 작성

token 을 만드는 방법, token을 생성할 때 담을 내용 결정

jaewoo9797 committed:
Create jwt token

클라이언트의 로그인 요청을 받아 유저 정보를 찾아와서 jwt 토큰을
생성하도록 만들었습니다.
수정해야 하는 부분, 컨트롤러에서 분리하여 책임을 분리한다

jaewoo9797 committed:
Refactor token, cookie 생성 로직 추상화

기존의 컨트롤러 내부에 너무 구체적인 내용들이 존재하여 읽기가 불편한
부분을 추상화하여 메소드로 분리했습니다

jaewoo9797 committed:
Refactor 기존의 유저를 찾고 토큰을 생성하는 로직을 서비스 레이어로 이관

컨트롤러에서 해당 작업을 처리하는 것이 적절하지 않다고 판단.

jaewoo9797 committed:
Refactor Member Dao 클래스 findByEmailAndPassword 메소드 구현로직 수정

기존의 queryForObject 예외가 발생했을 경우 예외처리를 위한 분기처리가
필요해서 스트림으로 처리할 수 있도록 query 메소드로 변경

jaewoo9797 committed:
Feat 토큰을 이용해서 유저 이름을 가져오는 API 구현 getAuthenticationInfo

요청에 담긴 쿠키에서 token 이름으로 토큰 값을 가지고와서 변환한 값으로
멤버의 정보를 조회해 응답에 맞는 객체로 변환하여 리턴하는 메소드 구현

jaewoo9797 committed:
사용하지 않는 의존성 삭제, 설정파일 yaml파일 형식에 맞게 수정,

leegwichan commented on src/main/java/roomescape/member/MemberController.java:
```diff
@@ -1,19 +1,19 @@
 @RestController
 public class MemberController {
     private MemberService memberService;
+    private static final String SECRET_KEY = "Yn2kjibddFAWtnPJ2AFlL8WXmohJMCvigQggaEypa5E=";
```
1. `SECRET_KEY`는 JWT를 생성할 때 언제 사용되나요?
2. 해당 값이 다른 사람들에게 노출되면 어떤 일이 일어날 수 있을까요?

jaewoo9797 replied:
1. 현재 서비스 레이어에서 JWT 토큰을 생성할 때와 파싱할 때 해당 `SECRET_KEY` 상수값을 이용해서 처리하고 있습니다. 초기에 컨트롤러에서 처리하던 로직이 서비스 레이어로 이관하면서 지우지 못하였네요.
2. 보안의 위험이 있습니다. 로직이 서비스로 이전되어 현재는 사용하지 않으니 컨트롤러에 존재하는 상수는 제거해두었습니다.

보안을 위해서 클래스 파일에 스트링 값을 하드코딩 하지 않고, application 설정 파일에 값을 두는 것으로 바꾸려고 합니다. 설정 파일에 위치를 이관시키고 생각해볼 점이 깃허브에 올라가지 않도록 gitignore를 시키는 것과, env 파일에 작성하고 환경변수에서 불러오는 것 중 선택을 해야하는데, 이 부분은 아직 잘 모르겠습니다.

그래서 application 파일에서 값을 가지고 오는 방식으로 변경했습니다.

leegwichan replied:
좋네요! application.yml을 어떻게 관리해야 github에 올리지 않게 할 수 있을지 알아보는 것도 좋을 것 같아요!
하지만 지금 미션에서는 중요한 부분은 아니라고 생각되니 시간 되시면 알아보시는 걸 추천드립니다.

leegwichan commented on src/main/java/roomescape/member/MemberDao.java:
```diff
@@ -28,7 +39,7 @@ public Member save(Member member) {
 
     public Member findByEmailAndPassword(String email, String password) {
-        return jdbcTemplate.queryForObject(
+        return jdbcTemplate.query(
```
1. `queryForObject()`를 `query()`로 바꾼 이유는 무엇인가요?
2. `findById()`와 달리 Optional 없이 null 을 return 하시는 이유는 무엇인가요?

jaewoo9797 replied:
1. queryForObject()를 query()로 바꾼 이유
![image](https://github.com/user-attachments/assets/401620ab-1e34-4d65-957b-22fa3f5a0525)

2. Optional을 이용하도록 코드 수정 하기 전에 PR을 올렸습니다. ㅎㅎ 
이 부분에 대해서 한 가지 더 생각하고 있는점이 존재합니다. 이메일과 패스워드를 이용해서 디비에 조회하고 있습니다. 그런데 꼭 비밀번호도 함께 조회를 해야하는지 고민중입니다.

```text
변경하려는 로직
1. 이메일로만 디비에서 해당 데이터를 조회해온다.
2. 가지고 온 데이터를 처리해서 입력받은 패스워드와 같은지 서버에서 검사한다.

왜 이런 생각을 했나요?
- 현재 비밀번호가 암호화가 안되고 있다. 암호화는 클라이언트의 정보를 보호하기 위해서 필수이다.
- 암호화된 비밀번호를 암호화 알고리즘을 이용해서 동일한지 처리해야한다.
```

leegwichan replied:
1. 왜 Stream으로 처리해야 해요? 이전 로직에서 `Optional.ofNullable(member)`로 반환해도 되지 않나요?
2. 좋은 의견이네요! 나중에 Password 암호화가 들어간다면 곰곰처럼 했을 것 같아요! 이번 미션은 "JWT를 이용한 인증/인가"가 메인 주제이기 때문에 시간에 여유가 있으실 때 해보면 좋을 것 같습니다!

jaewoo9797 replied:
### Optional.ofNullable, stream 의 findFirst()
`Optional.ofNullable(T value)`
값이 없다면 EMPTY `new Optional<>(null)` 을 반환해줍니다. 

stream 의 `findFirst()` 리턴 타입 `Optional<T>` 

```text
     * Returns an {@link Optional} describing some element of the stream, or an
     * empty {@code Optional} if the stream is empty.
```
두 기능은 값이 있다면 Optional에 값을 감싸서 반환하고 없다면 null을 Optional로 감싸서 반환하는 동일한 로직을 가지고 있습니다. 

두 방법 중 한가지를 선택해서 구현을 할 수 있는데, 만약 `Optional.ofNullable` 을 선택한다면
1. 반환된 리스트가 비어있는지 확인한다.
2. 반환된 리스트에서 0번 인덱스의 요소를 가지고온다.
3. Optional로 감싸서 반환한다.

이에 반해서 스트림을 이용하게 된다면 해당 로직 처리를 분기 처리 없이 간편하게 할 수 있습니다. 그래서 스트림을 선택했습니다.

### 변경하기 전
변경하기 전에는 `queryForObject` 를 통해서 리스트로 반환되는 것이 아니기 때문에 변경하기 전 로직에서는 `Optional.ofNullable` 를 이용하는게 적합하다고 생각합니다. 그런데 해당 로직은 0개 또는 1개 초과일 경우에 예외가 발생합니다. 예외처리도 해주어야하고, 예외 발생 시 Optional로 null을 감싸서 반환해주는 부분까지 리소스가 들어간다고 생각합니다. 아래 리뷰해주신 내용과 어느정도 연결이 되는 것 같습니다.

제가 더 생각해볼점이 있다면 알려주세요!

leegwichan replied:
> 1. 반환된 리스트가 비어있는지 확인한다.
> 2. 반환된 리스트에서 0번 인덱스의 요소를 가지고온다.
> 3. Optional로 감싸서 반환한다.

제가 말한 것은 이전에 쓰여있었던 코드 `jdbcTemplate.queryForObject()`를 사용한 후에 해당 값을 `Optional.ofNullable()`로 감싸는 것이 어떻냐는 이야기였어요. "stream을 사용하기 위해서 `jdbcTemplate.query()`를 사용했다"라는 말이 적당하지 않다고 생각했습니다. (물론 곰곰도 이해하고 아래에 적어주셨네요)

> 그런데 해당 로직은 0개 또는 1개 초과일 경우에 예외가 발생합니다. 예외처리도 해주어야하고, 예외 발생 시 Optional로 null을 감싸서 반환해주는 부분까지 리소스가 들어간다고 생각합니다.

0개 일 때 `Optional.empty()`를 반환하는 방법도 있겠죠? 이 방법은 생각해보셨나요?
```java
private Optional<Reservation> findReservationById(Long id) {
        try {
            return Optional.ofNullable(jdbcTemplate.queryForObject("...", rowMapper, id));
        } catch (EmptyResultDataAccessException exception) {
            return Optional.empty();
        }
    }
```

leegwichan commented on src/main/java/roomescape/member/MemberService.java:
```diff
@@ -1,17 +1,64 @@
 
 @Service
 public class MemberService {
+    private static final String SECRET_KEY = "Yn2kjibddFAWtnPJ2AFlL8WXmohJMCvigQggaEypa5E=";
```
SECRET_KEY를 Controller와 Service에서 각각 관리하면 어떤 점이 불편할까요?

jaewoo9797 replied:
중복된 값이 존재하게 되고 변경이 필요할 경우 두 곳에서의 변경이 함께 이루어져야합니다.

이에 수정한 방법은
- 컨트롤러에서는 해당 변수를 더 이상 사용하지 않고 있엇음. -> 제거
- 환경 변수 값을 주입 받는 형식으로 변경

leegwichan commented on src/main/java/roomescape/member/MemberService.java:
```diff
@@ -1,17 +1,64 @@
 
     public MemberResponse createMember(MemberRequest memberRequest) {
-        Member member = memberDao.save(new Member(memberRequest.getName(), memberRequest.getEmail(), memberRequest.getPassword(), "USER"));
+        Member member = memberDao.save(
+                new Member(memberRequest.getName(), memberRequest.getEmail(), memberRequest.getPassword(), "USER"));
```
`"USER"`라는 텍스트 값을 서비스에 관리하는데, 역할이 늘어난다면 어떻게 관리하실 건가요?

jaewoo9797 replied:
해당 값은 이넘 타입으로 관리하는게 적절해보입니다. 상수로 관리되어야할 도메인 규칙이고 문자열 타입으로 관리할 경우에는 오타같은 휴먼 에러가 발생할 수 있기 때문이라고 생각하기 때문입니다.

leegwichan commented on src/main/java/roomescape/member/MemberService.java:
```diff
@@ -1,17 +1,64 @@
+        return new MemberResponse(member.getId(), member.getName(), member.getEmail());
+    }
+
+    private String createToken(MemberResponse memberResponse) {
+        return Jwts.builder()
+                .setSubject(memberResponse.getId().toString())
+                .claim("name", memberResponse.getName())
+                .claim("email", memberResponse.getEmail())
+                .signWith(Keys.hmacShaKeyFor(SECRET_KEY.getBytes()))
+                .compact();
+    }
```
1. JWT에 `name` 정보와 `email` 정보를 넣으면, JWT 만으로 해당 정보를 알 수 있나요?
2. JWT에는 어떤 정보들을 넣어야 할까요?

jaewoo9797 replied:
해당 내용은 
아래에서 주신 리뷰 내용을 먼저 해결한 후에 정리해보겠습니다.

leegwichan replied:
https://jwt.io/ 와 같은 웹 사이트에서 직접 JWT를 decode 해볼 수 있어요. 공부하실 때 참고해보세요~!

leegwichan commented on src/main/java/roomescape/member/MemberService.java:
```diff
@@ -1,17 +1,64 @@
+        return new MemberResponse(member.getId(), member.getName(), member.getEmail());
+    }
+
+    private String createToken(MemberResponse memberResponse) {
+        return Jwts.builder()
+                .setSubject(memberResponse.getId().toString())
+                .claim("name", memberResponse.getName())
+                .claim("email", memberResponse.getEmail())
+                .signWith(Keys.hmacShaKeyFor(SECRET_KEY.getBytes()))
+                .compact();
+    }
```
JWT 를 만드는 것, 정보를 가져오는 것은 Controller의 책임인가요? Service의 책임인가요? Repository의 책임인가요?

> - 컨트롤러는 웹 요청/응답 책임만 가지도록 수정해보세요.
> - 디비 접근 책임은 DAO(Data Access Object)가 가지도록 위임해보세요.
> - 비즈니스 플로우 책임은 서비스가 가지도록 위임해보세요.
> - 비즈니스 규칙 책임은 도메인이 가지도록 위임해보세요.

jaewoo9797 replied:
### JWT 생성과 변환은 무엇인가?

사용자가 로그인할 경우에 유저 정보를 토큰으로 만들어 서버에 요청 시마다 함께 제공하도록 만들고, 서버로 들어온 토큰이 유효한지 검사하는 것.

- 해당 로직은 비즈니스 플로우에서 책임을 가지고 처리되어야한다.

- 비즈니스 규칙이기 때문에 도메인으로 분리하여 책임을 위임해야한다.

- 분리된 도메인을 이용하여 비즈니스 플로우에서 JWT도메인을 이용하도록 한다.

생각을 정리해보고 생각대로 수정해보았습니다.

leegwichan replied:
곰곰이 생각하는 "비즈니스 규칙"이 무엇인가요?

jaewoo9797 replied:
느낌적인 느낌으로 알고 있었는데 이번 기회에 조금 명확하게 정리해보았습니다.

### 비즈니스 규칙과 플로우
앞서 예시를 통해서 설명해보겠습니다. `클라이언트의 회원가입` 에서는 
- 클라이언트로부터 서버의 요청 (컨트롤러)
- 기존에 같은 이름이 있는지 검사, 도메인 규칙에 맞는지 검사, 디비에 저장 요청 등의 플로우 (서비스 레이어)
- 이름은 10글자를 넘을 수 없다. (비즈니스 규칙)
- 데이터베이스에 저장 (데이터베이스)

<br/>

**비즈니스 플로우**
  - 애플리케이션의 전체적인 흐름을 다룬다.
  - 서비스 레이어에서 처리한다.
  - 클라이언트 요청을 받아서 필요한 **도메인 객체를 조합**하고, 여러 도메인 규칙을 실행하는 역할
  - 여러 **도메인 로직을 조립**하여 하나의 트랜잭션이나 기능을 완성하는 과정

**비즈니스 규칙**
  - 특정 도메인에 대한 규칙을 정의하는 역할
  - 특정한 객체가 자체적으로 지켜야 하는 규칙과 책임을 가진다.
  - 도메인 모델 또는 특정 책임을 가진 객체가 처리해야 한다.
  - 개별적인 단위의 로직을 갖고 있다.
  
즉 서비스레이어에서는 Jwt 토큰을 생성하거나 파싱을 하는 것이 아니라 전체적인 흐름을 관리하고 각 도메인이 가지는 규칙에 대한 것은 해당 객체에게 위임을 하는 것이라고 이해했습니다.

기존에 서비스 레이어에서 해도되지 않나? 라는 생각을 했었는데, 왜 JwtProvider 객체를 분리하여 위임하는지에 대해서 이해하게 된 것 같습니다.

Jwt 객체의 비즈니스 규칙
- jwt 를 생성하는 방식 (알고리즘, 서명방식, 클레임에 무엇을 포함할 것인가)
- jwt 를 파싱하는 방식 
- 비밀키 관리

leegwichan replied:
제가 생각하는 도메인과는 조금 다르네요. 현재 코드에서는 Member, Reservation, Time 과 같은 객체가 도메인이라고 생각되는데요. JWT와는 성격이 다른 것 같아요.

1. 곰곰이 생각하는 도메인은 무엇인가요?
2. 만약 JWT 방식에서 Session 방식으로 인가/인증 방식이 바뀐다면, Controller와 Service 중 어느 곳이 바뀌어야 할까요?

leegwichan commented on src/main/java/roomescape/member/MemberService.java:
```diff
@@ -1,17 +1,64 @@
+        return createToken(memberResponse);
+    }
+
+    private MemberResponse findByEmailAndPassword(String email, String password) {
+        Member member = memberDao.findByEmailAndPassword(email, password);
+        // TODO null check 하고 예외 반환 로직 필요
+        return new MemberResponse(member.getId(), member.getName(), member.getEmail());
+    }
```
이번 기회에 하시죠 ^^

leegwichan commented on src/test/java/roomescape/MissionStepTest.java:
테스트 작성 계획은 어떻게 되나요?

jaewoo9797 replied:
잘 모르겠습니다.. 모든 것을 테스트 할 수는 없으니, 선택을 해야하는 문제, 제한된 시간 안에 가능한가 에 대해서 고민입니다. 커찬이 최소한 테스트 코드를 작성하면서 가져갔으면 하는 부분이 존재할까요? 마음은 각 레이어와 생성한 객체 모두 테스트 코드 학습을 해보고 싶은데, 감당 못할 것 같습니다. 😢

leegwichan replied:
이번에 로그인을 구현했으니, Rest Assured를 이용한 로그인 API 테스트가 가장 우선순위가 높을 것 같네요.
만약 부담되신다면 다른 것부터 하시는 걸 추천드립니다!

leegwichan commented on src/main/java/roomescape/member/CheckResponse.java:
이 Response의 역할은 무엇인가요? (이름 만으로는 어디에 사용하는 Response인지 알기 어렵네요;;)

jaewoo9797 replied:
로그인한 유저의 이름을 가져오기 위해서 클라이언트의 요청에 대한 응답입니다.
CheckResponse 는 무엇을 체크한다는 것인지 의미가 불분명하여 `AuthUserNameResponse` 로 수정했습니다.

leegwichan commented on src/main/java/roomescape/member/MemberDao.java:
```diff
@@ -52,4 +63,9 @@ public Member findByName(String name) {
         );
     }
+
+    public Optional<Member> findById(Long id) {
+        String query = "SELECT id, name, email, role FROM member WHERE id = ?";
+        return jdbcTemplate.query(query, mapper, id).stream().findFirst();
+    }
```
해당 값이 여러 개인 경우는 어떻게 되나요?

jaewoo9797 replied:
`query` 메소드의 결과로 여러 값이 나올 수 있습니다. 
- 디비에 요청할 때 Unique 한 값으로 조회해오는가? -> pk 값은 유일하기 때문에 여러 개인 경우가 없다.
- 만약, 유니크하지 않은 값을 이용해서 조회해온다. -> 조회 조건을 단 건만 가지고 오도록 해야한다. 그렇지 않다면 여러 값이 조회된다. 그럴 경우에는 1개의 값만 가지고 오도록 조건을 명확하게 설정해주어야 한다. 즉, 조회 조건이 유일한 값을 이용해서 조회해오던가 여러 건을 가져와서 후처리를 더 해주어야 한다.

현재 id 로 가져오는 경우는 있거나, 없거나 둘 중 한 가지입니다. 만약 여러개가 나왔다면 디비의 pk 값이 설정이 잘 되어있는지 확인해야합니다. 

메소드의 목적에 맞게 사용합니다. 정확하게 단건 복수건을 분리합니다.

leegwichan replied:
> 디비에 요청할 때 Unique 한 값으로 조회해오는가? -> pk 값은 유일하기 때문에 여러 개인 경우가 없다.

해당 경우라면 `jdbcTemplate.query()` 보가는 `jdbcTemplate.queryForObject()`를 사용할 것 같은데 어떻게 생각하시나요?

jaewoo9797 replied:
저는 `queryForObject`를 사용할 경우에 예외처리를 해주어야 한다는 점에서 `query` 메소드를 선택하게 되었습니다. 두 메소드 모두 데이터베이스에서 해당 데이터를 찾아서 반환해주는 로직인데, 저는 예외처리를 하여 코드가 try catch 구문이 들어가서 가독성이 떨어지는 것 때문에 `query` 메소드를 선택했습니다.

leegwichan replied:
(개인적인 의견) try-catch가 그렇게까지 가독성이 안떨어진다고 생각하는데요. 만약 곰곰이 try-catch 문을 작성해보고 비교했을 때 가독성이 많이 떨어진다고 느꼈다면 곰곰의 의견을 존중하겠습니다.
개인적으로는 굳이 List로 받아서 stream 만드는 비용이 더 크다고 생각합니다.

leegwichan commented on src/main/java/roomescape/member/MemberController.java:
```diff
@@ -25,13 +25,26 @@ public ResponseEntity createMember(@RequestBody MemberRequest memberRequest) {
+        return ResponseEntity.ok().header(HttpHeaders.SET_COOKIE, cookie.toString()).build();
+    }
+
+    private ResponseCookie createCookie(String accessToken) {
+        return ResponseCookie.from("token", accessToken)
+                .path("/")
+                .httpOnly(true)
+                .build();
     }
+
```
`"token"`이라는 값도 상수로 관리해야 하는 것 아닌가요?

jaewoo9797 replied:
상수로 분리했습니다.

leegwichan reviewed (changes requested):
안녕하세요 곰곰! 커찬입니다.
구현해주신 코드에 대해 리뷰 남겼습니다! 리뷰 반영 & 코멘트 작성 하시고 다시 리뷰 요청주세요!

jaewoo9797 committed:
Refactor 코드 리뷰 내용 반영 수정

기존의 알기 힘든 CheckResponse 네이밍 변경 해당 객체의 목적인 로그인한 유저의 정보를 반환하는 것을 반영해 네이밍 변경
MemberController에 사용하지 않던 상수 제거,Service 레이어에 존재하는 secret key 환경변수에서 가져와서 초기화 하도록 변경
MemberRepository의 findByEmailAndPassword 반환 타입으로 Optional

jaewoo9797 committed:
Create JwtProvider, 기존 서비스 레이어 안에 해당 로직 객체로 이관

jaewoo9797 reviewed (commented):
커찬! 리뷰 주신 내용들을 수정해보았습니다. 답변을 받고 Jwt 에 대해서 조금 더 학습을 해보고 수정해보려고 합니다!

커밋 단위로 수정을 하였습니다. 이번에는 조금 못쪼갰는데, 다음부터는 하나의 커밋에 하나의 목적을 가지고 커밋을 해두겠습니다 ㅎㅎ..

leegwichan commented on src/main/java/roomescape/member/MemberController.java:
```diff
@@ -1,37 +1,52 @@
         return ResponseEntity.created(URI.create("/members/" + member.getId())).body(member);
     }
 
-    @PostMapping("/logout")
-    public ResponseEntity logout(HttpServletResponse response) {
-        Cookie cookie = new Cookie("token", "");
-        cookie.setHttpOnly(true);
-        cookie.setPath("/");
-        cookie.setMaxAge(0);
-        response.addCookie(cookie);
-        return ResponseEntity.ok().build();
```
로그아웃 기능을 지운 이유는 무엇인가요?

jaewoo9797 replied:
아마 토큰 관련 로직을 서비스 레이어로 옮기면서 지운것 같은데, 실수인것 같습니다.
`Refactor 기존의 유저를 찾고 토큰을 생성하는 로직을 서비스 레이어로 이관 백재우 25. 3. 7. 오후 4:12`
깃 커밋 내용 찾아보니까 실수로 함께 지워버린 것 같습니다. 다시 만들어 두었습니다.

leegwichan commented on src/main/java/roomescape/member/MemberService.java:
```diff
@@ -4,14 +4,56 @@
-    private MemberDao memberDao;
 
-    public MemberService(MemberDao memberDao) {
+    public static final String ROLE_VALUE = "USER";
```
ROLE_VALUE 라고 하기에는, 역할이 여러개 있어 적절하지 않아 보여요. 더 좋은 이름이 있지 않을까요?

jaewoo9797 replied:
음... `DEFAULT_ROLE` ?? 잘 모르겠습니다.

leegwichan replied:
현재 상황에서는 `DEFAULT_ROLE` 정도도 괜찮겠네요.

leegwichan commented on src/main/java/roomescape/member/MemberService.java:
```diff
@@ -4,14 +4,56 @@
+                new Member(memberRequest.getName(), memberRequest.getEmail(), memberRequest.getPassword(), ROLE_VALUE));
+    }
+
+    public String authenticateAndGetToken(LoginRequest loginRequest) {
+        MemberResponse memberResponse = findByEmailAndPassword(loginRequest.email(), loginRequest.password());
+        return jwtProvider.createToken(memberResponse);
+    }
```
1. authenticate와 authorize의 차이는 무엇인가요?
2. (개인적인 생각) "메서드는 한 가지 역할을 해야한다" 라는 점에서 위와 같은 경우에 `And` 라는 표현은 지양해야 한다고 생각해요. 곰곰은 어떻게 생각하시나요?

jaewoo9797 replied:
| |authenticate(인증) | authorization(인가)|
|--|--------------------|-----------------------|
|의미|본인이 누군지 확인한다.(로그인) | 특정 리소스에 권한이 있는지 확인한다.(등급 권한) |
|예| 클라이언트가 서비스에 등록된 고객인지 확인 | 클라이언트가 하고자 하는 작업이 허가된 작업인지 확인|

제가 생각하는 인증과 인가의 차이입니다. 

2. 맞는것 같습니다. 음... 하나의 메소드에 두 가지 일을 하면서 And 라는 접속사를 사용하게 되었네요. 생각을 해보니 해당 메소드의 목적은 클라이언트가 로그인을 하는 로직을 처리하고 있네요. 그리고 내부적으로는 아이디와 비밀번호로 사용자를 찾고 해당 유저 정보를 토큰을 생성하고 있네요.

그래서  `login` 이라는 메소드 네이밍이 적절할 것 같아서 수정해보았습니다.

leegwichan replied:
1. 만약에 기존의 이름 형식을 유지했다면 `authenticate` 보다는 `authorization`이 더 좋다고 생각해요!
2. (개인적인 의견) `login`이라는 이름도 나쁘진 않은 것 같아요. 하지만 반환 값이 String인데 이 값이 JWT 값이라고 바로 인지할 수 있을 지는 의문이네요.

leegwichan reviewed (changes requested):
안녕하세요 곰곰! 커찬입니다.
기존 코멘트에 대한 답변과 추가적인 질문을 남겼습니다. 확인하시고 리뷰 반영 & 코멘트 남겨주세요!

jaewoo9797 committed:
Refactor Member Controller 의 authenticateAndGetToken 메소드 명을 login 으로 수정

메소드가 수행하는 기능을 추상화하지 못하고 내부적인 구체적인 구현으로 네이밍하여 And 라는 접속사를 사용한 것을 수정했습니다. 메소드의 구현 로직이 수행하고자 하는 목적인 login이 적절하다고 판단하였습니다.

jaewoo9797 reviewed (commented):
커찬 수정할 부분은 수정하고 리뷰 주신것에 대한 생각을 답변했습니다.

이제, 
- 컨트롤러 API 테스트하는 방법에 대해서 알아보기

- Jwt 에는 무슨 값을 담아야하는가?
  - 클레임, 서브젝트

- 왜 Jwt 기능은 정적 메소드가 아닌 콤포넌트로 만들어 Spring container에서 관리하게 했는가?

에 대해서 정리하고 적용해보려고 합니다.

leegwichan reviewed (approved):
일단 1차적으로 이야기 하고 싶은 것은 이야기 한 것 같아 다음 미션으로 넘어가겠습니다.
아직 해결 못한 문제에 대해서는 다음 PR 제출하실 때 같이 첨부해주시면 좋을 것 같아요. (시간이 충분히 많다고 생각하고 질문을 좀 많이 드릴께요. 바쁘시면 위에서 부터 답변해주세요!)

1. JWT는 어떻게 구성되어 있는가? `claim`에 담긴 값을 볼 수 있는 대상은 누구인가?
2. JWT secret key가 외부에 노출 될 경우, 보안적으로 어떤 문제가 있는가?
3. JWT를 Member, Reservation, Time과 같은 도메인으로 볼 수 있을까? 곰곰이 생각하는 도메인이란 무엇인가? 
4. 만약 JWT 방식에서 Session 방식으로 인가/인증 방식이 바뀐다면, Controller와 Service 중 어느 곳이 바뀌어야 할까요?
5. Rest Assured 를 이용하여 구현한 API 테스트 작성
